### PR TITLE
initial import of BITFURY16 driver

### PR DESCRIPTION
**Pull request by hotmine**

### CODE REVIEW

File: bf16-bitfury16.c [hunk 1]
```diff
@@ -0,0 +1,1190 @@
...
+		command->data_length++;
+	}
+
+	command->depth = depth;
+
+	command->cmd_code = cmd_code;
+	command->tx[2] = command->cmd_code;
+	command->data_length++;
+
+	if ((data_length >=0) && (data_length <= 79)) {
```
Comment (ckolivas) [inline bf16-bitfury16.c]: bf16-bitfury16.c:729:19: warning: comparison is always true due to limited range of data type [-Wtype-limits]
  if ((data_length >=0) && (data_length <= 79)) {

File: bf16-bitfury16.c [hunk 2]
```diff
@@ -0,0 +1,1190 @@
...
+		list->count--;
+		free(wdata->data);
+		free(wdata);
+	} else
+		return -1;
+
+	return 0;
+}
+
+int8_t workd_list_remove(bf_list_t* list, struct cgpu_info *bitfury, bf_works_t* works) {
```
Comment (ckolivas) [inline bf16-bitfury16.c]: bf16-bitfury16.c:873:61: warning: unused parameter ‘bitfury’ [-Wunused-parameter]
 int8_t workd_list_remove(bf_list_t* list, struct cgpu_info *bitfury, bf_works_t* works) {

File: driver-bitfury16.c [hunk 3]
```diff
@@ -0,0 +1,4589 @@
...
+	end = strchr(start + 1, ']');
+	if (end == NULL)
+		return -1;
+	len = end - start;
+
+	memset(buff, 0, sizeof(buff));
+	cg_memcpy(buff, start + 1, len);
+	chip_id = atoi(buff);
+
+	if ((board_id < 0) || (board_id >= CHIPBOARD_NUM)) {
```
Comment (ckolivas) [inline driver-bitfury16.c]: driver-bitfury16.c:670:16: warning: comparison is always false due to limited range of data type [-Wtype-limits]
  if ((board_id < 0) || (board_id >= CHIPBOARD_NUM)) {

File: driver-bitfury16.c [hunk 4]
```diff
@@ -0,0 +1,4589 @@
...
+		applog(LOG_ERR, "%s: invalid board_id %d: [0 - %d] specified",
+				bitfury->drv->name,
+				board_id, CHIPBOARD_NUM);
+		return -1;
+	}
+
+	if ((bcm250_id < 0) || (bcm250_id >= BCM250_NUM)) {
```
Comment (ckolivas) [inline driver-bitfury16.c]: driver-bitfury16.c:677:17: warning: comparison is always false due to limited range of data type [-Wtype-limits]
  if ((bcm250_id < 0) || (bcm250_id >= BCM250_NUM)) {

File: driver-bitfury16.c [hunk 5]
```diff
@@ -0,0 +1,4589 @@
...
+				bitfury->drv->name,
+				board_id, bcm250_id, chip_id);
+
+		uint8_t toggle[4] = { 0xa5, 0x00, 0x00, 0x02 };
+
+		ret = cmd_buffer_push(cmd_buffer,
+				info->chipboard[board_id].bcm250[bcm250_id].channel_depth, chip_address, chip_address,
+				work, 0, CHIP_CMD_TOGGLE, 3, toggle);
+
+		if (ret < 0)
```
Comment (ckolivas) [inline driver-bitfury16.c]: driver-bitfury16.c:1495:11: warning: comparison is always false due to limited range of data type [-Wtype-limits]
   if (ret < 0)

File: driver-bitfury16.c [hunk 6]
```diff
@@ -0,0 +1,4589 @@
...
+		} else {
+			gen_clock_data(bf16_chip_clock, 1, clock_buf);
+			info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].clock = bf16_chip_clock;
+		}
+
+		ret = cmd_buffer_push(cmd_buffer,
+				info->chipboard[board_id].bcm250[bcm250_id].channel_depth, chip_address, chip_address,
+				work, 0, CHIP_CMD_SET_CLOCK, 3, clock_buf);
+
+		if (ret < 0)
```
Comment (ckolivas) [inline driver-bitfury16.c]: driver-bitfury16.c:1524:11: warning: comparison is always false due to limited range of data type [-Wtype-limits]
   if (ret < 0)

File: driver-bitfury16.c [hunk 7]
```diff
@@ -0,0 +1,4589 @@
...
+		memset(noncemask, 0, sizeof(noncemask));
+
+		for (i = 0; i < 4; i++)
+			noncemask[i] = (mask >> (8*(4 - i - 1))) & 0xff;
+
+		ret = cmd_buffer_push(cmd_buffer,
+				info->chipboard[board_id].bcm250[bcm250_id].channel_depth, chip_address, chip_address,
+				work, 0, CHIP_CMD_SET_MASK, 3, noncemask);
+
+		if (ret < 0)
```
Comment (ckolivas) [inline driver-bitfury16.c]: driver-bitfury16.c:1546:11: warning: comparison is always false due to limited range of data type [-Wtype-limits]
   if (ret < 0)

File: driver-bitfury16.c [hunk 8]
```diff
@@ -0,0 +1,4589 @@
...
+						info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].cwork.payload.nbits, mask,
+						info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].cwork.task);
+
+				ret = cmd_buffer_push(cmd_buffer,
+						info->chipboard[board_id].bcm250[bcm250_id].channel_depth, chip_address, chip_address,
+						info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].cwork, 0,
+						CHIP_CMD_TASK_WRITE, 79,
+						info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].cwork.task);
+
+				if (ret < 0)
```
Comment (ckolivas) [inline driver-bitfury16.c]: driver-bitfury16.c:1583:13: warning: comparison is always false due to limited range of data type [-Wtype-limits]
     if (ret < 0)

File: driver-bitfury16.c [hunk 9]
```diff
@@ -0,0 +1,4589 @@
...
+					board_id, bcm250_id, chip_id);
+
+			ret = cmd_buffer_push(cmd_buffer,
+					info->chipboard[board_id].bcm250[bcm250_id].channel_depth, chip_address, chip_address,
+					work, 0, CHIP_CMD_TASK_STATUS, 0, NULL);
+
+			info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].status_time.tv_sec  = curr_time.tv_sec;
+			info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].status_time.tv_usec = curr_time.tv_usec;
+
+			if (ret < 0)
```
Comment (ckolivas) [inline driver-bitfury16.c]: driver-bitfury16.c:1618:12: warning: comparison is always false due to limited range of data type [-Wtype-limits]
    if (ret < 0)

File: driver-bitfury16.c [hunk 10]
```diff
@@ -0,0 +1,4589 @@
...
+			(opt_bf16_renonce == RENONCE_DISABLED)) {
+			applog(LOG_DEBUG, "%s: chipworker_thr: chip [%d:%d:%2d], prepare read_nonce_cmd",
+					bitfury->drv->name,
+					board_id, bcm250_id, chip_id);
+
+			ret = cmd_buffer_push(cmd_buffer,
+					info->chipboard[board_id].bcm250[bcm250_id].channel_depth, chip_address, chip_address,
+					work, 0, CHIP_CMD_READ_NONCE, 0, NULL);
+
+			if (ret < 0)
```
Comment (ckolivas) [inline driver-bitfury16.c]: driver-bitfury16.c:1637:12: warning: comparison is always false due to limited range of data type [-Wtype-limits]
    if (ret < 0)

File: driver-bitfury16.c [hunk 11]
```diff
@@ -0,0 +1,4589 @@
...
+			uint8_t new_buff = ((cmd_status.status & 0x0f) == 0x0f) ? 1 : 0;
+
+			/* status cmd counter */
+			info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].status_cmd_dx++;
+			info->chipboard[board_id].bcm250[bcm250_id].status_cmd_dx++;
+			info->chipboard[board_id].status_cmd_dx++;
+			info->status_cmd_dx++;
+
+			/* check if chip task has switched */
+			if (new_buff != info->chipboard[board_id].bc
```
Comment (ckolivas) [inline driver-bitfury16.c]: driver-bitfury16.c:3218:23: warning: comparison is always true due to limited range of data type [-Wtype-limits]
        if ((bcm250_id >= 0) && (bcm250_id < BCM250_NUM / 2)) {

File: bf16-bitfury16.c [hunk 12]
```diff
@@ -0,0 +1,1190 @@
...
+				if (i == 48)
+					break;
+
+				nonce = 0x00000000;
+			}
+
+			nonce |= (command->rx[i + 2] << 8*(4 - (i%4) - 1));
+		}
+
+#if 0
```
Comment (ckolivas) [inline bf16-bitfury16.c]: Remove unused code like this if your driver is complete please.

File: bf16-bitfury16.c [hunk 13]
```diff
@@ -0,0 +1,1190 @@
...
+		.cmd_code        = CHIP_CMD_SET_MASK,
+		.cmd_description = "CHIP_CMD_SET_MASK"
+	},
+	{
+		.cmd_code        = CHIP_CMD_CREATE_CHANNEL,
+		.cmd_description = "CHIP_CMD_CREATE_CHANNEL"
+	}
+};
+
+char* get_cmd_description(bf_cmd_code_t cmd_code) {
```
Comment (ckolivas) [inline bf16-bitfury16.c]: Style niggle: Please use new line for function braces.

File: bf16-bitfury16.c [hunk 14]
```diff
@@ -0,0 +1,1190 @@
...
+}
+
+static void shift_bits(uint8_t* data, uint8_t size, uint8_t nbits) {
+	uint8_t i;
+	uint8_t bytes = nbits / 8;
+	uint8_t bits  = nbits % 8;
+
+	for (i = 0; i < size; i++) {
+		data[i] = (data[i + bytes] << bits);
+		uint8_t minor = (data[i + bytes + 1] >> (8 - bits));
```
Comment (ckolivas) [inline bf16-bitfury16.c]: Style niggle: Please don't mix declaratives and code.

File: bf16-bitfury16.c [hunk 15]
```diff
@@ -0,0 +1,1190 @@
...
+		cmd_status->status = command->rx[0];
+		command->checksum             = command->rx[1];
+		cmd_status->checksum_expected = command->checksum;
+
+		if (nonces == NULL)
+			return 1;
+
+		/* fill nonces buffer */
+		uint32_t nonce = 0x00000000;
+		for (i = 0; i <= 48; i++) {
```
Comment (ckolivas) [inline bf16-bitfury16.c]: What's the 48 magic number here? Consider a named macro.

File: bf16-bitfury16.h [hunk 16]
```diff
@@ -0,0 +1,382 @@
+#pragma once
+
+#include <stdint.h>
+
+#include "bf16-spidevice.h"
+
+#ifdef __cplusplus
```
Comment (ckolivas) [inline bf16-bitfury16.h]: Why are these here? Cgminer is pure c code only.

File: cgminer.c [hunk 17]
```diff
@@ -1588,6 +1597,48 @@ static struct opt_table opt_config_table[] = {
 		     set_int_0_to_100, opt_show_intval, &opt_bxm_bits,
 		     "Set BXM bits for overclocking"),
 #endif
+#ifdef USE_BITFURY16
+	OPT_WITHOUT_ARG("--bf16-set-clock",
+			opt_set_bool, &opt_bf16_set_clock,
```
Comment (ckolivas) [inline cgminer.c]: Build with other drivers fails here onwards with:
cgminer.c:1602:19: error: ‘opt_bf16_set_clock’ undeclared here (not in a function)
    opt_set_bool, &opt_bf16_set_clock,

with the following generic configure:
./configure --enable-avalon --enable-avalon2 --enable-avalon4 --enable-avalon7 --enable-bflsc --enable-bitforce --enable-bitfury --enable-bitfury16 --enable-blockerupter --enable-cointerra --enable-drillbit --enable-hashfast --enable-hashratio --enable-icarus --enable-klondike --enable-modminer

File: configure.ac [hunk 18]
```diff
@@ -282,6 +282,17 @@ if test "x$bitfury" = xyes; then
...
 
+bitfury16="no"
+
+AC_ARG_ENABLE([bitfury16],
+	[AC_HELP_STRING([--enable-bitfury16],[Compile support for BitFury 16nm ASICs (default disabled)])],
+	[bitfury16=$enableval]
+	)
+if test "x$bitfury16" = xyes; then
+	AC_DEFINE([USE_BITFURY16], [1], [Defined to 1 if BitFury 16nm ASIC support is wanted])
+	drivercount=x$drivercount
```
Comment (ckolivas) [inline configure.ac]: Should this device be labelled standalone here? I'm not sure whether the hardware can be generic with other hardware or not.

File: api.c [hunk 19]
```diff
@@ -27,7 +27,7 @@
 #include "klist.h"
 
 #if defined(USE_BFLSC) || defined(USE_AVALON) || defined(USE_AVALON2) || defined(USE_AVALON4) || \
-  defined(USE_HASHFAST) || defined(USE_BITFURY) || defined(USE_BLOCKERUPTER) || defined(USE_KLONDIKE) || \
+  defined(USE_HASHFAST) || defined(USE_BITFURY) || defined(USE_BITFURY16) || defined(USE_BLOCKERUPTER) || defined(USE_KLONDIKE) || \
```
Comment (ckolivas) [inline api.c]: Missing device code for bitfury16 in api.c further down.

File: configure.ac [hunk 20]
```diff
@@ -885,7 +902,7 @@ else
 fi
 
 #Add any new device to this, along with a no on the end of the test
-if test "x$avalon$avalon2$avalon4$avalon7$avalon_miner$bab$bflsc$bitforce$bitfury$blockerupter$hashfast$hashratio$icarus$klondike$knc$modminer$drillbit$minion$cointerra$bitmine_A1$ants1$ants2$ants3$sp10$sp30" = xnonononononononononononononononononononononononono; then
+if test "x$avalon$avalon2$avalon4$avalon7$avalon_miner$bab$bflsc$bitforce$bitfury$bifury16$blockerupter$hashfast$hashratio$icarus$klondike$knc$modminer$drillbit$minion$cointerra$bitmine_A1$ants1$ants2$ants3$sp10$sp30" = xnononononononononononononononononononononononononono; then
```
Comment (ckolivas) [inline configure.ac]: Misspelt your own device name.

File: driver-bitfury16.c (hunk 3 above)
Comment (hotmine) [inline driver-bitfury16.c]: this one is fixed

File: driver-bitfury16.c (hunk 4 above)
Comment (hotmine) [inline driver-bitfury16.c]: this one is fixed

File: driver-bitfury16.c (hunk 5 above)
Comment (hotmine) [inline driver-bitfury16.c]: this one is fixed

File: driver-bitfury16.c (hunk 6 above)
Comment (hotmine) [inline driver-bitfury16.c]: this one is fixed

File: driver-bitfury16.c (hunk 7 above)
Comment (hotmine) [inline driver-bitfury16.c]: this one is fixed

File: driver-bitfury16.c (hunk 8 above)
Comment (hotmine) [inline driver-bitfury16.c]: this one is fixed

File: driver-bitfury16.c (hunk 9 above)
Comment (hotmine) [inline driver-bitfury16.c]: this one is fixed

File: driver-bitfury16.c (hunk 10 above)
Comment (hotmine) [inline driver-bitfury16.c]: this one is fixed

File: bf16-bitfury16.c (hunk 12 above)
Comment (hotmine) [inline bf16-bitfury16.c]: I have left this code untouched as driver is still in development - and communication debug is sometimes needed. If you insist on removing this code - I will do this

File: bf16-bitfury16.c (hunk 14 above)
Comment (hotmine) [inline bf16-bitfury16.c]: This one is fixed
I did not check other code - perhaps some more revision is needed regarding styling...

File: cgminer.c (hunk 17 above)
Comment (hotmine) [inline cgminer.c]: this one is fixed

File: configure.ac (hunk 18 above)
Comment (hotmine) [inline configure.ac]: yep, you are right
I've marked my device as standalone, as it is run on specific hardware

File: api.c (hunk 19 above)
Comment (hotmine) [inline api.c]: it is not missing
it has define:

**Comment by hotmine** [comment]
Hello, Con
Thanks for your help and contribution.
I have created another pull request in order to merge all our preproduction
driver changes.
At the moment driver is almost done and gets its final changes. So there
will be no more big changes.
Please take a look at them when you have time.
Thanks in advance.


2017-03-25 3:07 GMT+02:00 Con Kolivas <notifications@github.com>:
